File: baekjoon/basic2/1992/1992.cpp
// 백준 1992 : 쿼드트리 (https://www.acmicpc.net/problem/1992)

#include <stdio.h>
#include <string>
using namespace std;
int a[65][65];
string ret;
int n;

string go(int y, int x, int n){
    string ret = "";

    if(n==1){
        ret += to_string(a[y][x]);
        return ret;
    } 
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            if(a[y+i][x+j] != a[y][x]){
                ret += ("("+go(y, x, n/2) + go(y, x+n/2, n/2) + go(y+n/2, x, n/2) + go(y+n/2, x+n/2, n/2)+")");
                return ret;
            } 
        }
    }
    ret += to_string(a[y][x]);
    return ret;
}

int main(){
    scanf("%d", &n);
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            scanf("%1d", &a[i][j]);
        }
    }
    ret = go(0, 0, n);
    printf("%s\n", ret.c_str());
    return 0;
}

File: baekjoon/basic2/1992/1992.py
# 백준 1992 : 쿼드트리 (https://www.acmicpc.net/problem/1992)

import sys
sys.setrecursionlimit(10**6)

def go(y, x, n):
    if n == 1: return str(a[y][x])
    ret = ""
    for i in range(y, y+n):
        for j in range(x, x+n):
            if a[i][j] != a[y][x]:
                ret += ("("+go(y, x, n//2)+go(y, x+n//2, n//2)+go(y+n//2, x, n//2)+go(y+n//2,x+n//2,n//2)+")")
                return ret
    return str(a[y][x])

n = int(sys.stdin.readline())
a = []
for i in range(n):
    a.append(list(map(int, sys.stdin.readline().rstrip())))
sys.stdout.write(go(0,0,n)+"\n")